build_tools/cmake/iree_filter_test.cmake

```diff
@@ -0,0 +1,72 @@
+# Copyright 2023 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+include(CMakeParseArguments)
+
+function(iree_filter_test)
+  cmake_parse_arguments(
+    _FUNC
+    ""
+    "RESULT_VAR_ENABLED;TIMEOUT;DRIVER"
+    "LABELS"
+    ${ARGN}
+  )
+
+  set(_ENABLED TRUE)
+
+  # Exclude hostonly tests when cross-compiling.
+  if(CMAKE_CROSSCOMPILING AND "hostonly" IN_LIST _FUNC_LABELS)
+    set(_ENABLED FALSE)
+  endif()
+
+  # Sanitizer-based exclusions.
+  # CUDA is considered incompatible with all sanitizers.
+  if(IREE_ENABLE_ASAN)
+    if (("noasan" IN_LIST _FUNC_LABELS) OR (_FUNC_DRIVER STREQUAL "cuda"))
+      set(_ENABLED FALSE)
+    endif()
+  endif()
+  if(IREE_ENABLE_TSAN)
+    if (("notsan" IN_LIST _FUNC_LABELS) OR (_FUNC_DRIVER STREQUAL "cuda"))
+      set(_ENABLED FALSE)
+    endif()
+  endif()
+
+  # Architecture-based exclusions.
+  if ((IREE_ARCH MATCHES "^riscv_") AND ("noriscv" IN_LIST _FUNC_LABELS))
+    set(_ENABLED FALSE)
+  endif()
+
+  # Timeout-based exclusions.
+  #
+  # Tests timeout values are "short", "moderate", "long". The IREE-wide default
+  # is "short".
+  # On very slow configurations, we only run "short" tests (the IREE-wide
+  # default), excluding "moderate" and "long". On other somewhat-slow configs,
+  # we run "short" and "moderate" tests, excluding "long".
+  set(_EXCLUDE_TIMEOUT_MODERATE OFF)
```

Comment:
Hmm I don't really like just not configuring these tests if someone sets ASAN or RISC-V. Like it's very reasonable that we choose not to run this on CI, but I don't think we should hard-prevent running a long test with ASAN or a moderate one with RISC-V. Real RISC-V hardware does exist. I feel a bit less strongly about things like disabling Cuda with ASAN, since it's actually just a hard failure (right?), but even there I think someone should be able to try running a cuda test with asan and see what happens (who knows, maybe they can fix whatever is wrong with cuinit).

That's why test exclusions on the CI are mostly based on test label filtering at `ctest` runtime. Of course, then people complain that just running `ctest .` fails and that is also a fair complaint... As usual we have a conflict between the desire for simple things to "just work" and general principles against being overly prescriptive. Again a potential place for CMake/CTest presets and a configure script. Ugh...

Sorry for all the bikeshedding on this. I appreciate you trying to clean this up.

Reply:
> Sorry for all the bikeshedding on this. I appreciate you trying to clean this up.

No, it is important that you don't let me make a change here that you're not fully on board with. Sorry about the dreaded point-by-point reply here, it gives the wrong impression. If this doesn't immediately clear things up, we can give a try to a video call whenever works for you, or I can just drop this altogether if the benefit is not worth the cost of your and Scott's time here.

> Hmm I don't really like just not configuring these tests if someone sets ASAN or RISC-V.

That's already the case. I'm only changing how we do that disabling --- more localized and explicit --- not the fact that this disabling of tests at configure-time is happening.

For example, a lot of longer-than-default-timeout tests are currently tagged `noriscv`, and I'm changing to instead tagging them more specifically about the longer-than-default-timeout, and then in this one place infer that that means disablign them on `riscv`.

> Real RISC-V hardware does exist.

Yes. As noted in the comment being added 2 lines below here,

    # We should have a dedicated
    # setting for emulators instead of treating all RISC-V as very slow.

I have now filed https://github.com/openxla/iree/issues/14197 so I will update this PR with a TODO linking to it.

> That's why test exclusions on the CI are mostly based on test label filtering at ctest runtime. Of course, then people complain that just running ctest . fails and that is also a fair complaint... 

There are two problems with treating this as `ctest` filtering.

- The problem you mention here, that `ctest` results in failure unless invoked exactly as in the CI scripts. +1 that the complaint about that is a fair one.
- The other problem, with bytecode-module-based tests such as `iree_check_test`, is that `ctest` filtering only saves the overhead of running CTest, not the overhead of building the bytecode modules (`iree-test-deps`). That's a significant problem: on CI, where ccache is well set up for the C/C++ building steps, by contrast the non-ccache'd `iree-test-deps` build is substantial overhead. And in sanitizer configs, the compiler itself is slower than normal, so for larger tests, the latency gets high --- exactly what we are trying to filter out here.

Reply:
>  but even there I think someone should be able to try running a cuda test with asan and see what happens (who knows, maybe they can fix whatever is wrong with cuinit

Brainstorming... what if we add more finer grained controls to `iree_filter_test`, so it wouldn't _just_ set `RESULT_VAR_ENABLED`, but would set `RESULT_VAR_DISABLED;RESULT_VAR_DEFINED`. "DEFINED" would determine whether to declare the targets/tests at all (current "ENABLED" behavior) and "DISABLED" would set the [`DISABLED` test property](https://cmake.org/cmake/help/latest/prop_test/DISABLED.html). If the property had a bypass mechanism (I don't see one), someone could then choose to run it anyway. Maybe we could add our own setting like `IREE_INCLUDE_DISABLED_TESTS` that would just leave the property off and let people run tests that _can configure_ but _are expected to fail_. I guess continuing that line of thought, we could also use the [`WILL_FAIL` test property](https://cmake.org/cmake/help/latest/prop_test/WILL_FAIL.html), if we want to run the tests but check that they fail (I'm not sure we'd want to run sanitizer tests in that way though... seems wasteful). Putting those ideas together, `IREE_INCLUDE_DISABLED_TESTS` could set `WILL_FAIL` _instead of_ `DISABLED`.